### ISSUE TITLE
[RFC]: Encoder/decoder models & feature compatibility

### ISSUE THREAD

**Issue by afeldman-nm** [vllm-project/vllm#7366]
## Motivation <a href="#user-content-motivation" id="motivation">#</a>

There is significant interest in vLLM supporting encoder/decoder models. Issues #187  and #180 , for example, request encoder/decoder model support. As a result encoder/decoder support was recently introduced to vLLM via the following three PRs:

* #4837 
* #4888 
* #4942 

These three PRs make encoder/decoder model inference possible; however, they leave more to be desired in terms of (1) parity between vLLM's decoder-only & encoder/decoder request processing pipelines with respect to feature support, and (2) the number of encoder/decoder models which are supported.

The ask for the vLLM community is to contribute PRs which help bring vLLM encoder/decoder functionality to a similar level of maturity as that of vLLM's decoder-only functionality.

## Proposed changes <a href="#user-content-proposed-changes" id="proposed-changes">#</a>

The support matrix below summarizes which encoder/decoder models have already been added & which features are currently compatible with the vLLM encoder/decoder pipeline, versus which features & models will require additional PRs to implement in the long-term:

<table>
  <tr>
    <th>Model/feature</th>
    <th>Model is already available/feature is already compatible with encoder-decoder?</th>
    <th>Having this model/making this feature compatible is a long-term goal?</th>
  </tr>
  <tr>
    <td>Encoder/decoder infrastructure</td>
    <td><strong><u>Yes</u></strong></td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>BART</td>
    <td><strong><u>Yes</u></strong></td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Whisper</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>T5</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Other enc/dec models</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Quantization</td>
    <td><strong><u>Untested</u></strong></td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Multimodality</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Attention backends other than Xformers (esp. flash-attn, flashinfer)</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Custom attention bias support</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>CUDAGraph</td>
    <td>No<br>(Issue #7447)</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Pipeline parallelism</td>
    <td>No</td>
    <td><strong><u>Yes</u></strong></td>
  </tr>
  <tr>
    <td>Speculative decoding</td>
    <td>No</td>
    <td><strong>Low-priority but nice-to-have; difficult.</strong></td>
  </tr>
  <tr>
    <td>Automatic prefix caching</td>
    <td>No</td>
    <td><strong>Low-priority; difficult.</strong></td>
  </tr>
  <tr>
    <td>Sliding window</td>
    <td>No</td>
    <td>No</td>
  </tr>
  <tr>
    <td>Chunked prefill</td>
    <td>No</td>
    <td>No</td>
  </tr>
  <tr>
    <td>LoRA</td>
    <td>No</td>
    <td>No</td>
  </tr>
</table>

This RFC gives an overview of those features & models which **are not compatible with encoder/decoder currently, but which should be made compatible eventually** (i.e. **No** in the second column, **Yes** in the third column in the support matrix.)

Note that there are features (automatic prefix caching/sliding window/chunked prefill/LoRA) which are not long-term compatibility goals.

## Background <a href="#user-content-background" id="background">#</a>

Before continuing, it will be helpful to review [the details of the new vLLM encoder/decoder infrastructure](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/infra-enc-dec.md). 

It will also be helpful to review [this how-to guide](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/how-to.md) for adding new encoder/decoder models & improving encoder/decoder feature compatibility.

## Initial goal <a href="#user-content-initial-goal" id="initial-goal">#</a>

Members of the vLLM contributor community identify models/features in the support matrix above, for which they will work on writing a PR.

### Add new models to vLLM <a href="#user-content-add-new-models-to-vllm" id="add-new-models-to-vllm">#</a>

Please review the [how-to guide for adding new models to vLLM](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/how-to.md#guide-to-adding-new-encoderdecoder-models-to-vllm)

See `tests/models/test_bart.py` for an example of an encoder/decoder model unit test. See `tests/distributed/test_basic_distributed_correctness_enc_dec.py` for an example of an encoder/decoder model test with TP > 1.

#### Add Whisper model <a href="#user-content-add-whisper-model" id="add-whisper-model">#</a>

Steps to add support for [Whisper](https://paperswithcode.com/paper/robust-speech-recognition-via-large-scale-1f), a multimodal encoder/decoder speech recognition model:
* [Extend existing vLLM multimodality support to encoder/decoder models](#support-encoderdecoder-multimodality)
* Extend existing vLLM prompt processing pipeline to support audio
* Port [HuggingFace Whisper model](https://github.com/huggingface/transformers/blob/main/src/transformers/models/whisper/modeling_whisper.py) to vLLM; an existing open PR for this workstream is #5964 
* Modify each Whisper layer, where appropriate, to support TP > 1
* Add a Whisper test under `tests/models/`

Proposal: consider whether or not it makes sense to implement encoder/decoder multimodality, audio support, and Whisper in the same PR; that way, the Whisper model may be used to facilitate an end-to-end test with of audio multimodality.

#### Add T5 model <a href="#user-content-add-t5-model" id="add-t5-model">#</a>

Note: T5 depends on [custom attention bias being supported](#support-custom-attention-bias) by at least one of the attention backends which [also supports encoder attention & cross-attention](#add-support-for-encoder-attention-and-cross-attention-to-additional-backends); at time of writing no vLLM attention backend fulfills this requirement. The vLLM XFormers attention backend is the only backend which supports encoder/decoder models but neither it nor any other vLLM attention backend supports custom attention bias. (Custom attention bias is required in order to support T5 [relative positional encoding.](#custom-attention-bias-and-relative-positional-encoding))

Steps to add support for the [T5 model](https://paperswithcode.com/method/t5):
* Port [HuggingFace T5 model](https://github.com/huggingface/transformers/blob/main/src/transformers/models/t5/modeling_t5.py) to vLLM
  * This includes porting over the method which computes the custom attention bias matrix for T5 relative position encoding
* Modify each T5 layer, where appropriate, to support TP > 1
  * The custom attention bias computation must also support TP > 1
* Add a T5 test to `tests/models/`

Note: T5 was added to an older version of vLLM in #3117 , which could be a helpful starting-point

#### Add other encoder/decoder models

* Review open vLLM issues on GitHub and identify other encoder/decoder models which are requested by users

### Quantization <a href="#user-content-quantization" id="quantization">#</a>

The goal of this workstream is to make sure that quantization + encoder/decoder models is fully-tested, and to fill in any gaps (should they exist) in vLLM's support for quantized encoder/decoder models.

Steps to ensure that vLLM supports encoder/decoder models in combination with all existing vLLM quantization methods:

* Identify the list of quantization methods which vLLM currently supports with decoder-only models.
* Add unit tests for encoder/decoder models with all of these quantization methods.
* Determine which quantization methods are currently incompatible with vLLM encoder/decoder infrastructure.
* Scope out the effort involved in making these quantization methods compatible & submit a PR making the change.

vLLM encoder/decoder infrastructure should be compatible with most of the existing vLLM quantization methods, because the specialized quantization kernels are only employed for GEMM operations involving the learned weight matrices ($W_q$, $W_k$, etc.), whereas the encoder/decoder work really only modifies how the `Attention(q, k, v, kv_cache)` layer behaves & does not impact the learned weight matrices at all.

It is less clear whether vLLM encoder/decoder infrastructure would be incompatible with FP8. It does appear that a specialized quantized KV cache kernel is employed by the `Attention(q, k, v, kv_cache)` layer when FP8 quantization is employed.

### Support encoder/decoder multimodality <a href="#user-content-support-encoderdecoder-multimodality" id="support-encoderdecoder-multimodality">#</a>

Technically, vLLM already supports multimodality for models which have an "encoder" and a "decoder", i.e. Llava. However, Llava's decoder does not utilize cross-attention & the model is basically compatible with vLLM's pre-existing decoder-only infrastructure.

But critically, for **encoder/decoder models with cross-attention** such as Whisper vLLM does not currently support multimodality of any sort. The processing pipeline does not extract or utilize multimodal data from the input prompt, and the `EncoderDecoderModelRunner` has an assert which fails if the multimodal config is not `None`. Addressing this is what is meant by "supporting encoder/decoder multimodality".

Steps to extend existing vLLM multimodality support to encoder/decoder models:
* Review [existing vLLM multimodality support in the decoder-only pipeline](https://docs.vllm.ai/en/latest/dev/multimodal/multimodal_index.html)
* Scope out a plan for adding encoder/decoder multimodality support.
* Propose & implement one or more multimodal prompt formats for encoder/decoder models
* Integrate multimodality support into encoder/decoder processing pipeline
* Remove the assertion which fails when multimodality is enabled for an encoder/decoder model (see `assert_enc_dec_mr_supported_scenario()` in `vllm/worker/utils.py`)
* Add one or more unit tests with multimodal data

There are a number of multimodal encoder/decoder models which will benefit from this feature. One possibility is to add multimodality support & a multimodal model such as [Whisper](#add-whisper-model) in the same PR, so that Whisper may be used to facilitate an end-to-end test with multimodality.

Another possibility is to implement multimodality support in its own PR.

#### Considerations for designing multimodal encoder/decoder prompt formats <a href="#user-content-considerations-for-designing-multimodal-encoderdecoder-prompt-formats" id="considerations-for-designing-multimodal-encoderdecoder-prompt-formats">#</a>

One approach to designing the vLLM multimodal encoder/decoder prompt formats, is to consider what we want the user experience to be for high-priority multimodal encoder/decoder models such as
* [Llama 3.1 multimodal](https://github.com/vllm-project/vllm/pull/7258#discussion_r1710915145)
* [Whisper](#add-whisper-model)

#### Initial proposal for multimodal encoder/decoder prompt formats

It may be helpful to review
* [The non-multimodal encoder/decoder prompt formats which are currently supported by vLLM](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/infra-enc-dec.md#supported-encoderdecoder-prompt-formats): singleton prompts (raw text prompt, `TextPrompt`, `TokensPrompt`) as well as `ExplicitEncoderDecoder` prompts
* The multimodal decoder-only prompt formats which are currently supported by vLLM; search for `multi_modal_data` [here](https://github.com/vllm-project/vllm/blob/main/vllm/inputs/data.py) and also review the [vLLM documentation on multimodality](https://docs.vllm.ai/en/latest/dev/multimodal/multimodal_index.html)

Generally speaking, in encoder/decoder models based on cross-attention, the non-text input modality is passed to the encoder as input. Conversely, any text prompt is typically passed to the decoder as a input prompt.

The following two encoder/decoder multimodal prompt formats are tentatively proposed:

* Singleton `TextPrompt` with `multi_modal_data` field
    * vLLM will extract the `multi_modal_data` and pass it to the encoder module
    * vLLM will extract the prompt text, tokenize it and pass the token-list to the *decoder* (note that this is the opposite of vLLM behavior for non-multimodal prompts, where the prompt text would be passed to the encoder.)

    For example passing the `TextPrompt` below to vLLM BART

    ```
    TextPrompt(
        'prompt': "The rain in spain falls mainly on the",
        'multi_modal_data': <multi modal data structure>
    )
    ```

    results in

    ```
    Encoder input: <multi modal data structure>
    Decoder prompt: "The rain in spain falls mainly on the"
    ```

* Singleton `TokensPrompt` with `multi_modal_data` field
    * vLLM will extract the `multi_modal_data` and pass it to the encoder module
    * vLLM will extract the token list and pass it unmodified to the *decoder* (note that this is the opposite of vLLM behavior for non-multimodal prompts, where the prompt tokens would be passed to the encoder.)

    For example passing the `TokensPrompt` below to vLLM BART

    ```
    TokensPrompt(
        'prompt_tokens': [2,0,171,5,2],
        'multi_modal_data': <multi modal data structure>
    )
    ```

    results in

    ```
    Encoder prompt: <multi modal data structure>
    Decoder prompt: [2,0,171,5,2]
    ```

It may also be worth considering whether or how to support
* `ExplicitEncoderDecoderPrompt`s with multimodality
* An input prompt format which encapsulates *only* multimodal encoder inputs, with no associated decoder text/tokens prompt (this would result in the decoder being passed a "default" or empty prompt.)

### Add support for encoder attention and cross-attention to additional backends <a href="#user-content-add-support-for-encoder-attention-and-cross-attention-to-additional-backends" id="add-support-for-encoder-attention-and-cross-attention-to-additional-backends">#</a>

At time of writing, XFormers is the only vLLM attention backend which supports encoder attention & cross-attention. 

The goal of this workstream would be to extend encoder attention & cross-attention support to additional backends, the highest-priority being flash-attention and flashinfer.

Reviewing [encoder attention and cross-attention support in the XFormers backend would be a good starting-point](https://github.com/vllm-project/vllm/blob/main/vllm/attention/backends/xformers.py) for extending support to other models.

For context on the requirements for a backend to support encoder and cross-attention, it may help to review the [encoder/decoder architecture](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/infra-enc-dec.md#encoderdecoder-architecture-diagram-prefill--and-decode-phase), the [way that attention masks are currently constructed in the XFormers backend](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/infra-enc-dec.md#default-encoderdecoder-attention-bias-or-mask), and the [recommended architecture for vLLM encoder/decoder models](https://github.com/afeldman-nm/vllm-enc-dec-rfc/blob/main/how-to.md#25-optional-but-strongly-recommended-implement-the-following-encoderdecoder-model-architecture).

A summary of the key changes required for an attention backend to support encoder attention and cross-attention:
* The backend's `AttentionMetadata` subclass must support fields for encoder sequence lengths, encoder sequence token count, cross-attention blocktables, and cross-attention slot mapping. XFormers examples:
  * [`AttentionMetadata` subclass' encoder field declarations](https://github.com/vllm-project/vllm/blob/67abdbb42fdbb59c274130368981c0d0ac3539e3/vllm/attention/backends/xformers.py#L127-L140)
  * [Handle encoder & cross-attention fields in `prefill_metadata()` method](https://github.com/vllm-project/vllm/blob/67abdbb42fdbb59c274130368981c0d0ac3539e3/vllm/attention/backends/xformers.py#L216-L221)
  * [Handle encoder & cross-attention fields in `decode_metadata()` method](https://github.com/vllm-project/vllm/blob/67abdbb42fdbb59c274130368981c0d0ac3539e3/vllm/attention/backends/xformers.py#L255-L260)
* The `forward()` method of the backend implementation must accept an `attn_type` argument of type `AttentionType`, which allows choosing between encoder attention, decoder attention, or encoder/decoder cross-attention. [XFormers example](https://github.com/vllm-project/vllm/blob/67abdbb42fdbb59c274130368981c0d0ac3539e3/vllm/attention/backends/xformers.py#L447)
* The backend implementation must recognize which option has been chosen for `attn_type`, and adjust accordingly in terms of (1) how it utilizes `attn_metadata` when invoking the attention kernels (review [XFormers `forward()`](https://github.com/vllm-project/vllm/blob/67abdbb42fdbb59c274130368981c0d0ac3539e3/vllm/attention/backends/xformers.py#L438) for context), and (2) the choice of causal or non-causal attention, as well the choice of attention mask shape ([XFormers example](https://github.com/vllm-project/vllm/blob/67abdbb42fdbb59c274130368981c0d0ac3539e3/vllm/attention/backends/xformers.py#L689-L714)).

#### Initial goals
* Identify the changes required to add encoder attention & cross-attention support to flash-attention and flashinfer
* PR the required changes
  * Remove/modify any asserts which fail if the vLLM attention backend is not XFormers
    * Currently, [the `__init__()` method of `EncoderDecoderModelRunner`](https://github.com/vllm-project/vllm/blob/b4e9528f9569d6eb8c29624771a4058fe794cb5a/vllm/worker/enc_dec_model_runner.py#L95) invokes a method `EncoderDecoderModelRunner._maybe_force_supported_attention_backend()` defined [here](https://github.com/vllm-project/vllm/blob/b4e9528f9569d6eb8c29624771a4058fe794cb5a/vllm/worker/enc_dec_model_runner.py#L112-L144) which (1) attempts to force encoder/decoder models to use XFormers attention backend, and (2) raises an exception if the user has overridden the attention backend to be anything other than XFormers. 

#### Long-term goals
* All vLLM attention backends support encoder attention and cross-attention

### Support custom attention bias <a href="#user-content-support-custom-attention-bias" id="support-custom-attention-bias">#</a>

Note: [T5](#add-t5-model) takes a dependency on custom attention bias. Custom attention bias is likely complex enough to merit its own PR.

Note: custom bias support was added to `PagedAttention` in an older version of vLLM as part of #3117 ; given changes in vLLM since then, additional work would be required to integrate this implementation.

#### Custom attention bias and relative positional encoding

Attention bias refers to adding a matrix $A$ to the scaled dot-product (SDP) attention scores matrix before performing softmax, as shown below:

$$
attn(Q,K,V,A) = softmax(\frac{Q K^T + A}{\sqrt{d}})V
$$

Here, *custom* attention bias is understood to mean that the vLLM attention backend allows $A$ to be an arbitrary matrix, provided the tensor dimensions are commensurate with the shape of the SDP attention scores matrix. This is in contrast to the existing vLLM attention backend implementations, which can only accommodate simple block-diagonal causal or non-causal masks which are uniformly either $0$ or $-\infty$. 

There are broadly two possible approaches to custom attention bias, which do not necessarily have to be mutually-exclusive:
* $A$ is a fully-materialized attention bias matrix passed to the attention backend
* $A$ is computed on-the-fly by the attention kernel, using an element-wise formula for the attention bias which is fused with the $Q K^T$ and $softmax$ computations

T5 employs custom attention bias in order to implement [relative positional encoding](https://jaketae.github.io/study/relative-positional-encoding/#bridging-shaw-and-huang), wherein pairwise positional relationships between tokens are represented by the bias matrix. The HuggingFace Transformers T5 implementation provides an example of [how the relative positional encoding matrix is computed](https://github.com/huggingface/transformers/blob/c1aa0edb48217f416f4bbe6e3a9db1500284513b/src/transformers/models/t5/modeling_t5.py#L428).

#### Existing attention bias support

*Currently, no vLLM attention backend fully supports passing in a custom attention bias*. This is primarily due to underlying kernel limitations. For example, the [xFormers `memory_efficient_attention_forward` kernel](https://facebookresearch.github.io/xformers/components/ops.html) is the only NVIDIA-GPU-oriented kernel which permits passing in an arbitrary PyTorch tensor as a materialized attention bias (via the `attn_bias` argument) (at time of writing I have not investigated if custom attention bias is supported by any of the kernels for AMD GPU, CPU, etc.) Regardless, vLLM only employs xFormers `memory_efficient_attention_forward` for prefill; to my knowledge, none of the decode-phase kernels employed by vLLM can accept an arbitrary tensor as a custom attention bias, making custom attention bias impossible to apply end-to-end for both prefill and decode under the current vLLM implementation.

In addition to lack of kernel-level support for custom attention bias, most vLLM backends also prevent passing a custom attention bias matrix to the underlying kernel. The exception is the XFormers backend, which accepts an attention bias via `XFormersMetadata.attn_bias` attribute (however the XFormers backend only utilizes `attn_bias` in the prefill phase.)

#### Proposed methods for supporting custom attention bias

Here the following two approaches for supporting custom attention bias in vLLM are proposed:
* **Fully-materialized bias matrix:** Modify vLLM attention backends to accept an arbitrary PyTorch tensor, passed into the backend via the `AttentionMetadata.attn_bias` field.
* **On-the-fly/fused bias matrix computation:** Enable an efficient workflow whereby vLLM developers can tweak an attention kernel to compute the custom attention bias on the fly
  * For example: rather than computing the T5 relative position encoder bias matrix once, instead the attention kernel can fuse the element-wise bias matrix formula with the $Q K^T$ and $softmax()$. The attention bias matrix is never fully materialized.
  * [FlexAttention](https://pytorch.org/blog/flexattention/#relative-position-encodings) enables fused custom attention bias computations in a FlashAttention-style kernel, using torch.compile.

![image](https://pytorch.org/assets/images/flexattention/fg4.png)

It may make sense to support one or both of these methods.

Note that custom attention bias support must be added on a backend-by-backend basis, because of the kernel modifications & backend logic changes required.

#### Initial goals for introducing custom attention bias support

1. Focus on a particular vLLM attention backend
  * Suggestion: focus on an attention backend which also supports encoder/decoder models, in order to facilitate [running T5](#add-t5-model). At time of writing, XFormers is the only backend which supports encoder/decoder models, however there will likely be work on [supporting encoder/decoder in additional attention backends.](#add-support-for-encoder-attention-and-cross-attention-to-additional-backends)
2. Scope out the effort involved in introducing custom attention bias support to this backend
3. Some steps which will likely be involved in introducing custom attention bias support:
  * Augment attention backend's kernels to accept custom attention bias; for example, the PagedAttention kernel (for XFormers backend), the Flash-attention kernel (for the flash-attn backend), or the Flashinfer kernels (for the Flashinfer backend)
  * (Except for XFormers) add an `attn_bias` attribute to attention backend's `AttentionMetadata` subclass
  * Ensure that the attention backend passes the `attn_bias` attribute to both the prefill and decode kernels
4. Add at least two custom attention bias unit tests (for prefill & decode respectively)

#### Final goals for introducing custom attention bias support

* All vLLM attention backends should support custom attention bias, with unit tests

#### Some links which may be helpful for understanding how causal & non-causal attention masks are currently configured in vLLM:

* [Invocation of flash-attention for prefill in vLLM backend, using `causal` flag](https://github.com/vllm-project/vllm/blob/db35186391a2abfc6c91d703527dac20d2488107/vllm/attention/backends/flash_attn.py#L522)

* [Invocation of xFormers attention kernel for prefill in vLLM backend, using `BlockDiagonalMask` and `BlockDiagonalCausalMask`](https://github.com/vllm-project/vllm/blob/db35186391a2abfc6c91d703527dac20d2488107/vllm/attention/backends/xformers.py#L689-L738)

* [Invocation of FlashInfer attention kernel for prefill in backend, using `causal` flag](https://github.com/vllm-project/vllm/blob/db35186391a2abfc6c91d703527dac20d2488107/vllm/attention/backends/flashinfer.py#L539)

* [Invocation of PagedAttention kernel for decode in vLLM backend](https://github.com/vllm-project/vllm/blob/db35186391a2abfc6c91d703527dac20d2488107/vllm/attention/backends/xformers.py#L628)

* [Invocation of FlashInfer kernel for decode in vLLM backend](https://github.com/vllm-project/vllm/blob/db35186391a2abfc6c91d703527dac20d2488107/vllm/attention/backends/flashinfer.py#L543)

### Support CUDAGraph with encoder/decoder models <a href="#user-content-support-cudagraph-with-encoderdecoder-models" id="support-cudagraph-with-encoderdecoder-models">#</a>

Note: this topic is being tracked by Issue #7447

Steps to support CUDAGraph with encoder/decoder models:
* Scope out the effort require to support CUDAGraph with encoder/decoder models
* Write a PR for CUDAGraph + encoder/decoder
  * Remove the assertion which fails when CUDAGraph is enabled for an encoder/decoder model (see `assert_enc_dec_mr_supported_scenario()` in `vllm/worker/utils.py`)

### Support pipeline-parallelism with encoder/decoder models <a href="#user-content-support-pipeline-parallelism-with-encoderdecoder-models" id="support-pipeline-parallelism-with-encoderdecoder-models">#</a>

Steps to support pipeline-parallelism with encoder/decoder models:
* Scope out the effort required to support pipeline-parallelism with encoder/decoder models
* Write a PR for pipeline-parallelism + encoder/decoder
  * Remove the assertion which fails when pipeline-parallelism is enabled for an encoder/decoder model (see `assert_enc_dec_mr_supported_scenario()` in `vllm/worker/utils.py`)

### Support multi-step scheduling with encoder/decoder models <a href="#user-content-support-multi-step-scheduling-with-encoder-decoder-models" id="support-multi-step-scheduling-with-encoder-decoder-models">#</a>

Note: depends on #7000 landing in order to add multi-step scheduling support; it may be helpful to review the multi-step scheduling RFC ( #6854 )

Steps to support multi-step scheduling with encoder/decoder models:
* Scope out the effort required to support multi-step scheduling
  * `EncoderDecoderModelRunner` multi-step support
* Write a PR for multi-step scheduling + encoder/decoder
* Write at least one test of an encoder/decoder model with multi-step scheduling

### Low-priority high-effort tasks <a href="#user-content-low-priority-high-effort-tasks" id="low-priority-high-effort-tasks">#</a>

* Speculative decoding
* Automatic prefix caching

Here it is proposed that these features are low-priority. Adding support for speculative decoder and automatic prefix caching would require a significant of effort to scope out and design the implementations.

Note that adding support for either of these features would require removing the assertions which fail when speculative decoding or automatic prefix caching are enabled for an encoder/decoder model (see `assert_enc_dec_mr_supported_scenario()` in `vllm/worker/utils.py`)

### Feedback Period.

Closed.

### CC List.

@WoosukKwon 
@robertgshaw2-neuralmagic 
@mgoin 
@tms
@njhill 
@sroy745 
@ywang96 
@DarkLight1337 
@js8544

**Comment by afeldman-nm** [comment]
Looks like footnote references (i.e. [^1], [^2], etc.) are not rendered for RFCs on github, so I just edited the RFC to replace all of the footnote references with direct links.

**Comment by robertgshaw2-neuralmagic** [comment]
cc @mgoin

**Comment by afeldman-nm** [comment]
FYI, added a section to the RFC about adding multi-step scheduling + encoder/decoder support.

@robertgshaw2-neuralmagic

**Comment by DarkLight1337** [comment]
Sorry for mentioning this so late. For multimodal models, there are actually two ways to apply cross-attention:

- Cross-attention between text and multimodal features (e.g. Llama 3.1 multimodal)
- Cross-attention between text features only (i.e. multimodal encoder with an encoder-decoder language model) (e.g. #5934)

I wonder how the current plan could handle the latter case. To keep the API consistent, we should distinguish between the above two cases internally when multimodal data is passed.

**Comment by DarkLight1337** [comment]
Tagging #8811 for future reference

**Comment by NickLucche** [comment]
I can start looking into T5 support starting from the custom attention bias 🤞🏻

**Comment by robertgshaw2-neuralmagic** [comment]
> I can start looking into T5 support starting from the custom attention bias 🤞🏻

Thanks @NickLucche - this would be greatly appreciated

**Comment by sroy745** [comment]
@

> > I can start looking into T5 support starting from the custom attention bias 🤞🏻
> 
> Thanks @NickLucche - this would be greatly appreciated

Hi @NickLucche are you looking at any particular kernel for adding the custom bias needed for T5? FYI I am trying to add support to make the encoder-decoder models work with the flash_attn kernel.

**Comment by robertgshaw2-neuralmagic** [comment]
> @
> 
> > > I can start looking into T5 support starting from the custom attention bias 🤞🏻
> > 
> > 
> > Thanks @NickLucche - this would be greatly appreciated
> 
> Hi @NickLucche are you looking at any particular kernel for adding the custom bias needed for T5? FYI I am trying to add support to make the encoder-decoder models work with the flash_attn kernel.

@sroy745 - I think it’s probably easiest to do it with the paged attention backend, but doing in flash_attn would be the best

**Comment by njhill** [comment]
For us the most important is having something functional for T5 so was also thinking xformers option might be preferable/easier for the initial support.

**Comment by NickLucche** [comment]
@sroy745 @robertgshaw2-neuralmagic @njhill so xformers appears to be indeed the easiest, but it has to be paired with a `PagedAttention` modification too, AFAIK xformers is not used in decode because it has no concept of kv paging/block table.

FlashAttention would be super interesting to explore, but will likely require PRs to https://github.com/vllm-project/flash-attention or to the original repo. Even then I am not sure if a fully materialized bias matrix option would be accepted there, while the fused approach would be too specific to T5 to live there. 

Personally I would like to start with the "less-efficient"/easier approach first (fully materialized bias) and leave optimization for later, unless there's a more mature implementation we can somehow integrate here.

**Comment by jeejeelee** [comment]
Why isn't LoRA included as an LTS  feature? What considerations are there behind this decision? I've recently been working on supporting LoRA for `mllama` , and currently my main challenge is dealing with cross attention

**Comment by amybachir** [comment]
Is this feature still being planned to be supported? If so, any idea when it is planned for?

**Comment by DarkLight1337** [comment]
We already support encoder-decoder models, but it's currently limited to V0 only. We plan to eventually extend the support to V1 Engine.

**Comment by amybachir** [comment]
> We already support encoder-decoder models, but it's currently limited to V0 only. We plan to eventually extend the support to V1 Engine.

Thanks! I'm looking to serve T5. Looks like it isn't supported just yet!

**Comment by HaukurPall** [comment]
Is there an open PR for enabling support for encoder-decoder models in V1? I'm interested in (By)T5 and might be able to assist given some support.

**Comment by DarkLight1337** [comment]
We have decided to pull support for now because it adds too much to code complexity. cc @WoosukKwon
